url2df0f7c811_przemekgalazka_legacycode_tgz-pJ8-pl_cc_core_cmd_CommandJ8/src/pl/cc/Ping.java

```diff
@@ -0,0 +1,202 @@
+package pl.cc;
+
+import java.util.Date;
+
+import org.apache.log4j.Logger;
+
+import pl.cc.core.NetPrintWriter;
+import pl.cc.core.cmd.Command;
+import pl.cc.core.cmd.ErrSameAgentConnection;
+import pl.cc.core.cmd.InfoType;
+import pl.cc.core.cmd.InfoVersion;
+import pl.cc.core.cmd.OkWelcome;
+import pl.cc.utils.Utils;
+
+/**
+ * Wďż˝tek odpowiedzialny za podtrzymywanie komunikacji z agentem
+ * @since 2007-06-19
+ */
+public class Ping extends Thread implements ProxyEventListener {
+	static final Logger log = Logger.getLogger(Ping.class);
+	boolean stop = false;
+
+	ProxyEventListenerList proxyEventListenerList;
+	/**
+	 * Wďż˝tek sprawdzajďż˝cy pingi
+	 */
+	CheckPingThread checkPingThread = null;
+	
+	NetPrintWriter out;
+	/**
+	 * Czas ostatniego Ponga
+	 */
+	Date lastPong = Utils.getNow();
+
+	/**
+	 * Czas ostatniego Pinga
+	 */
+	Date lastPing;
+
+	int noPongCount = 0;
+	/**
+	 * ile sekund pomiďż˝dzy kolejnymi pingami
+	 */
+	int interval = 5; 
+	/**
+	 * ile sekund czekamy na ponga
+	 */
+	int timeOut = 10;
+
+
+	public Ping(ProxyEventListenerList proxyEventListenerList, NetPrintWriter out){
+		super();
+		this.proxyEventListenerList = proxyEventListenerList;
+		this.out = out;
+	}
+	
+	/**
+	 * @param proxy obiekt proxy
+	 * @param interval czas w sekundach pomiďż˝dzy kolejnymi pingami
+	 * @param timeOut tyle sek moze upďż˝ynďż˝ďż˝ od wpďż˝yniďż˝cia ostatniego ponga
+	 */
+	public Ping(ProxyEventListenerList proxyEventListenerList, NetPrintWriter out, int interval, int timeOut){
+		this(proxyEventListenerList, out);
+		this.interval = interval;
+		this.timeOut = timeOut;
+	}
+	
+	private void checkPong(){
+		if ((Utils.getNow().getTime()-lastPong.getTime())>(timeOut*1000)){
+			log.error("Brak ďż˝ďż˝cznoďż˝ci z ProxyAgent");
+			proxyEventListenerList.onProblemWithPing();
+			stop = true;
+		}
+	}
+
+	public void run(){
+		setName("ping");
+		log.info("Ping thread started");
+		checkPingThread = new CheckPingThread();
+		checkPingThread.start();
+		while (!stop){
+			try {
+				lastPing = Utils.getNow();
+				log.debug("Send PING: "+lastPing.toString());
+				out.println("ping");
+				sleep(interval*1000);
+				if (stop) break;
+			} catch (NullPointerException e1) {
+				log.warn("Problem z ping-pong (NullPointerException)");
+			} catch (InterruptedException e2) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)